src/guidellm/scheduler/constraints/over_saturation.py

```diff
@@ -0,0 +1,467 @@
+"""
+Over-saturation detection constraint implementation.
```

Comment:
Can we provide a short writeup of the method/algorithm here for what's implemented down below and how the classes interface?